### PR TITLE
fix : option value in dark mode #3801

### PR DESCRIPTION
**Pull request by rituraj2000**
### Feature Preview

Preview : Before | After

This update addresses the issue highlighted in [Issue #3801](https://github.com/AppFlowy-IO/AppFlowy/issues/3801).

fixes https://github.com/AppFlowy-IO/AppFlowy/issues/3801

---

#### PR Checklist

- [x] My code adheres to [AppFlowy's Conventions](https://docs.appflowy.io/docs/documentation/software-contributions/conventions).
- [x] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [x] All existing tests are passing.

### CODE REVIEW

**Comment by rituraj2000** [comment]
@annieappflowy Please review this PR and suggest any changes or improvements if necessary.

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 1]
```diff
@@ -68,25 +68,30 @@ class SelectOptionTag extends StatelessWidget {
...
-  }) : super(key: key);
+  final bool isTextColAlt;
+  const SelectOptionTag(
+      {required this.name,
+      required this.color,
+      this.onSelected,
+      this.onRemove,
+      this.isTextColAlt = false,
+      Key? key,
+      ac})
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:79]: This doesn't look right?

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 2]
```diff
@@ -68,25 +68,30 @@ class SelectOptionTag extends StatelessWidget {
...
-    Key? key,
-  }) : super(key: key);
+  final bool isTextColAlt;
+  const SelectOptionTag(
+      {required this.name,
+      required this.color,
+      this.onSelected,
+      this.onRemove,
+      this.isTextColAlt = false,
+      Key? key,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:78]: Use `super.key` constructor instead.

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 3]
```diff
@@ -112,7 +117,11 @@ class SelectOptionTag extends StatelessWidget {
               name,
               fontSize: FontSizes.s11,
               overflow: TextOverflow.ellipsis,
-              color: AFThemeExtension.of(context).textColor,
+              color: isTextColAlt
+                  ? Theme.of(context)
+                      .extension<AFThemeExtension>()!
+                      .textColorAlt
+                  : Theme.of(context).extension<AFThemeExtension>()!.textColor,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:124]:
```suggestion
              color: isTextColAlt
                  ? AFThemeExtension.of(context).textColorAlt
                  : AFThemeExtension.of(context).textColor,
```

File: frontend/appflowy_flutter/pubspec.lock
Comment (Xazin) [inline frontend/appflowy_flutter/pubspec.lock:1]: Why all the changes in this file?

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart (hunk 3 above)
Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:124]: Is `isTextColAlt` always true? If yes then why do we need it, and if no, then what am I missing?

Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:124]: If it's always true, can we calculate the contrast of the default `textColor` and use `textColorAlt` if the contrast is lower than some bound?

Comment (rituraj2000) [reply frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart]: @Xazin please review my latest changes. I have implemented this logic

**Comment by LucasXu0** [comment]
Hi, @rituraj2000. Please sign the CLA and update your code according to the suggestions.

**Comment by rituraj2000** [comment]
@Xazin I did the necessary changes. can you see it please

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart (hunk 1 above)
Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:79]: You still haven't resolved this.

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart (hunk 2 above)
Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:78]: Nor this

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 4]
```diff
@@ -98,6 +102,36 @@ class SelectOptionTag extends StatelessWidget {
...
+      // Ensure luminance1 is the lighter color
+      if (luminance1 < luminance2) {
+        double temp = luminance1;
+        luminance1 = luminance2;
+        luminance2 = temp;
+      }
+
+      // Calculate the contrast ratio
+      return (luminance1 + 0.05) / (luminance2 + 0.05);
+    }
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:133]: You should create a class for this, and abstract all magic numbers to constants with good names.

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 5]
```diff
@@ -112,7 +146,17 @@ class SelectOptionTag extends StatelessWidget {
...
+                        color,
+                        Theme.of(context)
+                            .extension<AFThemeExtension>()!
+                            .textColor,
+                      ) <
+                      1.36)
+                  ? Theme.of(context)
+                      .extension<AFThemeExtension>()!
+                      .textColorAlt
+                  : Theme.of(context).extension<AFThemeExtension>()!.textColor,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:159]: This should be done by a method, consider adding a helper method to the util class (see above comment).

That takes backgroundColor, primaryTextColor, and altTextColor as the arguments, and return the proper text color.

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/select_option_editor.dart [hunk 6]
```diff
@@ -259,6 +259,7 @@ class _CreateOptionCell extends StatelessWidget {
                 child: SelectOptionTag(
                   name: name,
                   color: AFThemeExtension.of(context).greyHover,
+                  isTextColAlt: true,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/select_option_editor.dart:262]: This seems redundant now

File: frontend/appflowy_flutter/packages/flowy_infra/lib/colorscheme/colorscheme.dart [hunk 7]
```diff
@@ -74,6 +74,7 @@ class FlowyColorScheme {
   final Color topbarBg;
   final Color icon;
   final Color text;
+  final Color textAlt;
```
Comment (Xazin) [inline frontend/appflowy_flutter/packages/flowy_infra/lib/colorscheme/colorscheme.dart:77]: This seems redundant now

File: frontend/appflowy_flutter/packages/flowy_infra/lib/colorscheme/colorscheme.dart [hunk 8]
```diff
@@ -128,6 +129,7 @@ class FlowyColorScheme {
     required this.topbarBg,
     required this.icon,
     required this.text,
+    required this.textAlt,
```
Comment (Xazin) [inline frontend/appflowy_flutter/packages/flowy_infra/lib/colorscheme/colorscheme.dart:132]: This seems redundant now

File: frontend/appflowy_flutter/pubspec.lock
Comment (Xazin) [inline frontend/appflowy_flutter/pubspec.lock:1]: You still haven't undone changes to this file

Comment (rituraj2000) [reply frontend/appflowy_flutter/pubspec.lock:1]: @Xazin resolved this now. please go through it once...

File: frontend/appflowy_flutter/packages/flowy_infra/lib/colorscheme/colorscheme.dart (hunk 8 above)
Comment (rituraj2000) [reply frontend/appflowy_flutter/packages/flowy_infra/lib/colorscheme/colorscheme.dart:132]: I tried for directly choosing black or white colour according to contrast but it did not give satisfactory results in all themes and lightings. So i added manually two texts themes : Dark and White colour.

**Comment by rituraj2000** [comment]
@Xazin @LucasXu0 Please review my latest PR. I am eager to take any further suggestions on changes.

**Comment by LucasXu0** [comment]
@rituraj2000 Have you been able to run this pull request on your local machine successfully?


Screenshot 2023-11-13 at 10 27 49

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 9]
```diff
@@ -108,12 +110,17 @@ class SelectOptionTag extends StatelessWidget {
...
+                fontSize: FontSizes.s11,
+                overflow: TextOverflow.ellipsis,
+                color: ColorAnalyzer().getAppropriateTextColor(
+                    color,
+                    Theme.of(context)
+                        .extension<AFThemeExtension>()!
+                        .textColorAlt,
+                    Theme.of(context)
+                        .extension<AFThemeExtension>()!
+                        .textColor)),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:123]: I already mentioned that you should use `AFThemeExtension.of(context)`, please correct.

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 10]
```diff
@@ -68,13 +71,12 @@ class SelectOptionTag extends StatelessWidget {
...
-    this.onSelected,
-    this.onRemove,
-    Key? key,
-  }) : super(key: key);
+  const SelectOptionTag({required this.name,
+      required this.color,
+      this.onSelected,
+      this.onRemove,
+      Key? key,
+      ac}): super(key: key);
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:79]: This still doesn't look right.

File: frontend/appflowy_flutter/lib/util/color_analyzer/color_analyzer.dart [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      double temp = luminance1;
+      luminance1 = luminance2;
+      luminance2 = temp;
+    }
+
+    return (luminance1 + 0.05) / (luminance2 + 0.05);
+  }
+
+  Color getAppropriateTextColor(
+      Color backgroundColor, Color primaryTextColor, Color altTextColor) {
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/util/color_analyzer/color_analyzer.dart:40]: Formatting, please add a commma after `altTextColor`.

File: frontend/appflowy_flutter/lib/util/color_analyzer/color_analyzer.dart [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    final double contrastRatio =
+        calculateContrast(backgroundColor, primaryTextColor);
+
+    return contrastRatio < 1.36 ? altTextColor : primaryTextColor;
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/util/color_analyzer/color_analyzer.dart:44]: What is this magic number `1.36`?

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart (hunk 2 above)
Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:78]: You still haven't resolved this issue.

**Review by Xazin — CHANGES_REQUESTED** [review]

**Comment by rituraj2000** [comment]
@Xazin @LucasXu0 Done with all the modifications you suggested. The 1.36 threshold value is also set as a constant and can be modified if better values are found. The : super (key, key) line was also fixed. 

Please do suggest if any more changes are needed.

File: frontend/appflowy_flutter/lib/util/color_analyzer/color_analyzer.dart [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    if (luminance1 < luminance2) {
+      final double temp = luminance1;
+      luminance1 = luminance2;
+      luminance2 = temp;
+    }
+
+    return (luminance1 + 0.05) / (luminance2 + 0.05);
+  }
+
+  Color getAppropriateTextColor(
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/util/color_analyzer/color_analyzer.dart:40]:
```suggestion
  static Color getAppropriateTextColor(
```

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 14]
```diff
@@ -112,7 +113,11 @@ class SelectOptionTag extends StatelessWidget {
               name,
               fontSize: FontSizes.s11,
               overflow: TextOverflow.ellipsis,
-              color: AFThemeExtension.of(context).textColor,
+              color: ColorAnalyzer().getAppropriateTextColor(
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:116]:
```suggestion
              color: ColorAnalyzer.getAppropriateTextColor(
```

File: frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart [hunk 15]
```diff
@@ -112,7 +113,11 @@ class SelectOptionTag extends StatelessWidget {
...
+                color,
+                Theme.of(context).extension<AFThemeExtension>()!.textColor,
+                Theme.of(context).extension<AFThemeExtension>()!.textColorAlt,
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:119]: You're still not using `AFThemeExtension.of(context)`, it's my third time mentioning it..
```suggestion
                AFThemeExtension.of(context).textColor,
                AFThemeExtension.of(context).textColorAlt,
```

Comment (rituraj2000) [reply frontend/appflowy_flutter/lib/plugins/database_view/widgets/row/cells/select_option_cell/extension.dart:119]: Oh..so sorry for this. i must have left it in the final commit...

**Comment by rituraj2000** [comment]
@Xazin @LucasXu0 @richardshiue could you please check my latest changes?

File: frontend/appflowy_flutter/lib/util/color_analyzer/color_analyzer.dart [hunk 16]
```diff
@@ -0,0 +1,50 @@
+import 'dart:math';
+import 'package:flutter/material.dart';
+
+class ColorAnalyzer {
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/util/color_analyzer/color_analyzer.dart:4]: Final thing, can you add a unit test to this location:
`frontend/appflowy_flutter/test/unit_test/color_analyzer_test.dart`

Where you test it with some predefined colors.

Let me know

**Comment by rituraj2000** [comment]
@Xazin @LucasXu0 @annieappflowy Please have a look at the PR and suggest if anymore changes are needed. I would love to move forward with more contributions to Appflowy.

**Comment by rituraj2000** [comment]
@Xazin Completed the changes. Please review it. Looking forward to more reviews !!

**Review by Xazin — APPROVED** [review]

**Comment by rituraj2000** [comment]
@LucasXu0 Is the PR causing some issue while merging ??? Should i work on it ??

**Review by skydiver1980 — APPROVED** [review]

**Comment by annieappflowy** [comment]
@LucasXu0 @Xazin good to merge?

**Comment by richardshiue** [comment]
Closing this as tests failed and an alternative design was implemented

**Pull request closed without merging**